### ISSUE TITLE
[Train] Worker fault tolerance recovery can result in gaps in reported metric logs

### ISSUE THREAD

**Issue by justinvyu** [ray-project/ray#41668]
### What happened + What you expected to happen

If you have worker fault tolerance enabled in Ray Train/Tune, it's possible to discard reported results (from `ray.train.report`), resulting in an incomplete result history at the end of the run.

Here's what can happen:
1. `storage_path` is set, which causes there to be a local "staging" version of the metrics logfiles (`result.json`) that gets uploaded to the storage path at some time-based frequency.
2. If a node goes down and `FailureConfig` is configured to try recovering the run, then the workers will be re-scheduled and the "trial" will restart (from the latest available checkpoint).
3. We currently have logic in the `json`/`csv` logger callbacks to pull the latest version of the file from `storage_path` to account for the "restoration from a new ray cluster / head node" situation. Whenever a trial "starts", then we'll download the latest version of the trial's `result.json` file to the local staging directory.
4. **The problem is that this logic also runs in the worker failure recovery case, which happens during the same job in the same ray cluster. So, you could have a local staging version that is way fresher than the one in `storage_path`, but it'll be overwritten by the stale one when we do this download.**

See the example below.

### Versions / Dependencies

2.7, 2.8, 2.9

### Reproduction script

In this example:
* `num_to_keep=2` is needed to force a driver sync of these metrics files to `storage_path` when the first 2 checkpoints get reported.
* Then, we report a bunch of metrics without a checkpoint, which doesn't force a sync of the files (at this point, the storage path is stale when it comes to these logfiles).
* Finally, we restore the run, which pulls these stale files down and overwrites the fresh files. This causes the end history to have a gap.

```python
import ray
from ray.train.data_parallel_trainer import DataParallelTrainer
from ray.train import ScalingConfig, RunConfig, FailureConfig

from ray.train.tests.util import create_dict_checkpoint, load_dict_checkpoint


def f(config):
    checkpoint = ray.train.get_checkpoint()
    start = 0
    if checkpoint:
        start = load_dict_checkpoint(checkpoint)["iter"] + 1

    for i in range(start, 30):
        if 0 <= i <= 1 or i > 20:
            with create_dict_checkpoint({"iter": i}) as checkpoint:
                ray.train.report({"iter": i}, checkpoint=checkpoint)
            if i == 21:
                raise RuntimeError("repro")
        elif 2 <= i <= 20:
            ray.train.report({"iter": i})


trainer = DataParallelTrainer(
    f,
    scaling_config=ScalingConfig(num_workers=2),
    run_config=RunConfig(
        failure_config=FailureConfig(max_failures=1),
        storage_path="/tmp/ray_results",  # must be specified for repro
        checkpoint_config=ray.train.CheckpointConfig(num_to_keep=2),
        # ^this forces a sync to happen so we have *something* stale uploaded to storage_path.
    ),
)
result = trainer.fit()

>>> result.metrics_dataframe["iter"]
0      0
1      1
2      2
3      3
4     22
5     23
6     24
7     25
8     26
9     27
10    28
11    29
Name: iter, dtype: int64
```

### Issue Severity

Low: It annoys or frustrates me.
